wire-ios-data-model/Source/Model/Conversation/ZMConversation+Federation.swift

```diff
@@ -29,4 +29,30 @@ extension ZMConversation {
         return domain != userDomain
     }
 
+    public static func groupConversationOwned(by domain: String, in context: NSManagedObjectContext) -> [ZMConversation]? {
+        let hostedOnDomainPredicate = NSPredicate(format: "%K == %d AND %K == %@",
+                                                  ZMConversationConversationTypeKey,
+                                                  ZMConversationType.group.rawValue,
+                                                  ZMConversationDomainKey,
+                                                  domain)
+        let request = self.sortedFetchRequest(with: hostedOnDomainPredicate)
+
+        return context.fetchOrAssert(request: request) as? [ZMConversation]
+    }
+
+    public static func groupConversationNotOwned(by domains: [String], in context: NSManagedObjectContext) -> [ZMConversation]? {
+        let groupConversationPredicate = NSPredicate(format: "%K == %d",
+                                                  ZMConversationConversationTypeKey,
+                                                  ZMConversationType.group.rawValue)
+        let hostedOnDomainPredicate = NSPredicate(format: "%K IN %@",
+                                                  ZMConversationDomainKey,
+                                                  domains)
+        let notHostedOnDomainPredicate = NSCompoundPredicate(notPredicateWithSubpredicate: hostedOnDomainPredicate)
+        let resultPredicate =  NSCompoundPredicate(andPredicateWithSubpredicates: [groupConversationPredicate, notHostedOnDomainPredicate])
+
+        let request = self.sortedFetchRequest(with: resultPredicate)
+
+        return context.fetchOrAssert(request: request) as? [ZMConversation]
+    }
+
```

Comment:
Suggestion to make this more readable and more reusable:

```swift
public extension ZMConversation {

    static func groupConversations(
        hostedOnDomain domain: String,
        in context: NSManagedObjectContext
    ) -> [ZMConversation] {
        let predicate: NSPredicate = .isGroupConversation.and(.isHostedOnDomain(domain))
        let request = sortedFetchRequest(with: predicate)
        return context.fetchOrAssert(request: request) as? [ZMConversation] ?? []
    }

    static func groupConversations(
        notHostedOnDomains domains: [String],
        in context: NSManagedObjectContext
    ) -> [ZMConversation] {
        let predicate: NSPredicate = .isGroupConversation.and(.isNotHostedOnDomains(domains))
        let request = sortedFetchRequest(with: predicate)
        return context.fetchOrAssert(request: request) as? [ZMConversation] ?? []
    }

}

private extension NSPredicate {

    static var isGroupConversation: NSPredicate {
        return hasConversationType(.group)
    }

    static func hasConversationType(_ type: ZMConversationType) -> NSPredicate {
        return NSPredicate(
            format: "%K == %d",
            ZMConversationConversationTypeKey,
            type.rawValue
        )
    }

    static func isHostedOnDomain(_ domain: String) -> NSPredicate {
        return NSPredicate(
            format: "%K IN %@",
            ZMConversationDomainKey,
            [domain]
        )
    }

    static func isNotHostedOnDomains(_ domains: [String]) -> NSPredicate {
        return isHostedOnAnyDomain(domains).inverse
    }


    static func isHostedOnAnyDomain(_ domains: [String]) -> NSPredicate {
        return NSPredicate(
            format: "%K IN %@",
            ZMConversationDomainKey,
            domains
        )
    }

}


public extension NSPredicate {

    var inverse: NSPredicate {
        return NSCompoundPredicate(notPredicateWithSubpredicate: self)
    }

    func and(_ other: NSPredicate) -> NSPredicate {
        return NSCompoundPredicate(andPredicateWithSubpredicates: [other])
    }

    func or(_ other: NSPredicate) -> NSPredicate {
        return NSCompoundPredicate(orPredicateWithSubpredicates: [other])
    }

}

```

Reply:
Implemented